snippets/base/models.py

```diff
@@ -602,6 +607,46 @@ def snippets(self):
 
         return ASRSnippet.objects.filter(pk__in=all_snippets).distinct()
 
+    def clean(self):
+        super().clean()
+
+        # If not new file, just return
+        if not isinstance(self.image.file, InMemoryUploadedFile):
+            return
+
+        im = Image.open(self.image.open())
+        if im.format != 'PNG':
+            raise ValidationError({'image': 'Upload only PNG images.'})
+
+        if (settings.IMAGE_MAX_DIMENSION and
+            (im.width > settings.IMAGE_MAX_DIMENSION or
+             im.height > settings.IMAGE_MAX_DIMENSION)):
+            raise ValidationError({
+                'image': 'Upload an image at most {0}x{0}. This image is {1}x{2}.'.format(
+                    settings.IMAGE_MAX_DIMENSION, im.width, im.height)
+            })
+
+        # Optimize only when settings is True
+        if settings.IMAGE_OPTIMIZE:
+            self.image.seek(0)
+            cmd = subprocess.run(
+                ['pngquant', '-', '--quality=95', '--skip-if-larger', '--speed=1'],
+                input=self.image.read(),
+                stdout=subprocess.PIPE
+            )
+            if cmd.stdout:
+                new_image = io.BytesIO(cmd.stdout)
+                self.image.file = InMemoryUploadedFile(
+                    new_image, 'ImageField', self.image.name,
+                    'image/png', len(new_image.read()), None
+                )
+
+        if settings.IMAGE_MAX_SIZE > 0 and self.image.size > settings.IMAGE_MAX_SIZE:
```

Comment:
is this max size checking the new size or the uploaded size? I see above that optimize has updated `self.image.file`, but I'm not sure if that updates the other metadata like `self.image.size`. I'm just wondering if the size check should come before the optimization step so that work can be avoided for files that are too large.

Reply:
Good catch. Actually `self.image.size` points to `self.image.file.size` in this case ([django code](https://github.com/django/django/blob/044cc544203fdcf2601ff200547ecae2e51a5db5/django/db/models/fields/files.py#L64)) so we're good. 

I want the size check after the optimization so things are simpler for the Editors. All we need is the final file size to be within limits, either pre-optimized or optimized by the django app.